File: backend/src/main/java/com/dynamicdashboard/cockpit/query/application/mapper/QueryMapper.java
package com.dynamicdashboard.cockpit.query.application.mapper;

import com.dynamicdashboard.cockpit.dashboard.repository.WidgetRepository;
import com.dynamicdashboard.cockpit.query.application.dto.QueryConditionDto;
import com.dynamicdashboard.cockpit.query.application.dto.QueryJoinDto;
import com.dynamicdashboard.cockpit.query.application.dto.QueryResponseDto;
import com.dynamicdashboard.cockpit.query.application.dto.QuerySortDto;
import com.dynamicdashboard.cockpit.query.application.dto.QueryTransformationDto;
import com.dynamicdashboard.cockpit.query.domain.DataQueryEntity;
import com.dynamicdashboard.cockpit.query.repository.QueryConditionRepository;
import com.dynamicdashboard.cockpit.query.repository.QueryGroupByFieldRepository;
import com.dynamicdashboard.cockpit.query.repository.QueryJoinRepository;
import com.dynamicdashboard.cockpit.query.repository.QuerySelectedFieldRepository;
import com.dynamicdashboard.cockpit.query.repository.QuerySortRepository;
import com.dynamicdashboard.cockpit.query.repository.QuerySourceBindingRepository;
import com.dynamicdashboard.cockpit.query.repository.QueryTransformationRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor
public class QueryMapper {

    private final QuerySourceBindingRepository querySourceBindingRepository;
    private final QuerySelectedFieldRepository querySelectedFieldRepository;
    private final QueryGroupByFieldRepository queryGroupByFieldRepository;
    private final QueryJoinRepository queryJoinRepository;
    private final QueryConditionRepository queryConditionRepository;
    private final QueryTransformationRepository queryTransformationRepository;
    private final QuerySortRepository querySortRepository;
    private final WidgetRepository widgetRepository;

    public QueryResponseDto toDto(DataQueryEntity entity) {
        if (entity == null) return null;

        List<String> sourceIds = querySourceBindingRepository.findByQueryIdOrderByPositionIndexAsc(entity.getId())
                .stream()
                .map(sb -> sb.getDataSource().getSourceKey())
                .collect(Collectors.toList());

        List<String> selectedFieldIds = querySelectedFieldRepository.findByIdQueryIdOrderByPositionIndexAsc(entity.getId())
                .stream()
                .map(sf -> sf.getField().getFieldKey())
                .collect(Collectors.toList());

        List<String> groupByFieldIds = queryGroupByFieldRepository.findByIdQueryIdOrderByPositionIndexAsc(entity.getId())
                .stream()
                .map(gb -> gb.getField().getFieldKey())
                .collect(Collectors.toList());

        List<QueryJoinDto> joins = queryJoinRepository.findByQueryId(entity.getId())
                .stream()
                .map(j -> QueryJoinDto.builder()
                        .id(j.getId().toString())
                        .leftSourceId(j.getLeftSource().getSourceKey())
                        .leftFieldId(j.getLeftField().getFieldKey())
                        .type(j.getJoinType().name().toLowerCase())
                        .rightSourceId(j.getRightSource().getSourceKey())
                        .rightFieldId(j.getRightField().getFieldKey())
                        .build())
                .collect(Collectors.toList());

        List<QueryConditionDto> conditions = queryConditionRepository.findByQueryId(entity.getId())
                .stream()
                .map(c -> QueryConditionDto.builder()
                        .id(c.getId().toString())
                        .fieldId(c.getField().getFieldKey())
                        .operator(c.getOperator().name().toLowerCase())
                        .value(c.getValueExpression())
                        .logical(c.getLogicalOperator().name())
                        .parametrable(c.isParameterizable())
                        .build())
                .collect(Collectors.toList());

        List<QueryTransformationDto> transformations = queryTransformationRepository.findByQueryId(entity.getId())
                .stream()
                .map(t -> QueryTransformationDto.builder()
                        .id(t.getId().toString())
                        .type(t.getTransformationType().name().toLowerCase())
                        .fieldId(t.getTargetField() != null ? t.getTargetField().getFieldKey() : null)
                        .outputLabel(t.getOutputLabel())
                        .formula(t.getFormulaExpression())
                        .format(t.getFormatOption())
                        .replacementValue(t.getReplacementValue())
                        .operator(t.getOperator() != null ? t.getOperator().name().toLowerCase() : null)
                        .value(t.getComparisonValue())
                        .build())
                .collect(Collectors.toList());

        QuerySortDto sortDto = querySortRepository.findById(entity.getId())
                .map(s -> QuerySortDto.builder()
                        .fieldId(s.getField().getFieldKey())
                        .direction(s.getDirection().name().toLowerCase())
                        .build())
                .orElse(null);

        int usedByWidgetsCount = widgetRepository.countByQueryId(entity.getId());

        return QueryResponseDto.builder()
                .id(entity.getId())
                .name(entity.getQueryName())
                .description(entity.getQueryDescription())
                .visibility(entity.getVisibility() != null ? entity.getVisibility().name().toLowerCase() : "shared")
                .sourceIds(sourceIds)
                .joins(joins)
                .selectedFieldIds(selectedFieldIds)
                .conditions(conditions)
                .transformations(transformations)
                .groupByFieldIds(groupByFieldIds)
                .aggregation(entity.getAggregation() != null ? entity.getAggregation().name().toLowerCase() : "none")
                .aggregationFieldId(entity.getAggregationField() != null ? entity.getAggregationField().getFieldKey() : null)
                .sort(sortDto)
                .rowLimit(entity.getRowLimit())
                .usedByWidgets(usedByWidgetsCount)
                .updatedAt(entity.getUpdatedAt())
                .build();
    }
}

File: frontend/src/app/core/api/mappers/query.mapper.ts
import { QueryResponseDto, QueryRequestDto, QueryJoinDto, QueryConditionDto, QueryTransformationDto, QuerySortDto } from '../dtos/query.dto';
import { DataQuery, QueryJoin, QueryCondition, QueryTransformation, QuerySort } from '@core/models/types';

export class QueryMapper {
  static toDomain(dto: QueryResponseDto): DataQuery {
    return {
      id: String(dto.id),
      name: dto.name,
      description: dto.description || '',
      visibility: dto.visibility ? (dto.visibility.toLowerCase() as any) : 'personal',
      sourceIds: dto.sourceIds || [],
      joins: (dto.joins || []).map(j => this.toJoinDomain(j)),
      selectedFieldIds: dto.selectedFieldIds || [],
      conditions: (dto.conditions || []).map(c => this.toConditionDomain(c)),
      groupByFieldIds: dto.groupByFieldIds || [],
      aggregation: dto.aggregation ? (dto.aggregation.toLowerCase() as any) : 'none',
      aggregationFieldId: dto.aggregationFieldId,
      sort: dto.sort ? this.toSortDomain(dto.sort) : (dto.sorts && dto.sorts.length > 0 ? this.toSortDomain(dto.sorts[0]) : undefined),
      transformations: (dto.transformations || []).map(t => this.toTransformationDomain(t)),
      rowLimit: dto.rowLimit || 1000,
      usedByWidgets: dto.usedByWidgets || 0,
      updatedAt: dto.updatedAt ? String(dto.updatedAt) : new Date().toISOString()
    };
  }

  static toDto(domain: DataQuery): QueryRequestDto {
    return {
      id: domain.id,
      name: domain.name,
      description: domain.description,
      visibility: domain.visibility ? domain.visibility.toUpperCase() : 'PERSONAL',
      sourceIds: domain.sourceIds || [],
      joins: (domain.joins || []).map(j => this.toJoinDto(j)),
      selectedFieldIds: domain.selectedFieldIds || [],
      conditions: (domain.conditions || []).map(c => this.toConditionDto(c)),
      groupByFieldIds: domain.groupByFieldIds || [],
      aggregation: domain.aggregation ? domain.aggregation.toUpperCase() : 'NONE',
      aggregationFieldId: domain.aggregationFieldId,
      sort: domain.sort ? this.toSortDto(domain.sort) : undefined,
      sorts: domain.sort ? [this.toSortDto(domain.sort)] : [],
      transformations: (domain.transformations || []).map(t => this.toTransformationDto(t)),
      rowLimit: domain.rowLimit || 1000,
      tags: [],
      usedByWidgets: domain.usedByWidgets,
      createdAt: domain.updatedAt,
      updatedAt: domain.updatedAt
    };
  }

  private static toJoinDomain(dto: QueryJoinDto): QueryJoin {
    return { id: dto.id, type: dto.type as any, leftSourceId: dto.leftSourceId, leftFieldId: dto.leftFieldId, rightSourceId: dto.rightSourceId, rightFieldId: dto.rightFieldId };
  }
  private static toJoinDto(domain: QueryJoin): QueryJoinDto {
    return { id: domain.id, type: domain.type, leftSourceId: domain.leftSourceId, leftFieldId: domain.leftFieldId, rightSourceId: domain.rightSourceId, rightFieldId: domain.rightFieldId };
  }

  private static toConditionDomain(dto: QueryConditionDto): QueryCondition {
    return { id: dto.id, fieldId: dto.fieldId, operator: dto.operator as any, value: dto.value, logical: (dto.logicalOperator as any) || 'AND', parametrable: false };
  }
  private static toConditionDto(domain: QueryCondition): QueryConditionDto {
    return { id: domain.id, fieldId: domain.fieldId, operator: domain.operator, value: domain.value, logicalOperator: domain.logical };
  }

  private static toSortDomain(dto: QuerySortDto): QuerySort {
    return { fieldId: dto.fieldId, direction: dto.direction as any };
  }
  private static toSortDto(domain: QuerySort): QuerySortDto {
    return { fieldId: domain.fieldId, direction: domain.direction };
  }

  private static toTransformationDomain(dto: QueryTransformationDto): QueryTransformation {
    return {
      id: dto.id,
      type: dto.type as any,
      fieldId: dto.fieldId,
      outputLabel: dto.outputLabel,
      formula: dto.formula,
      format: dto.format as any,
      replacementValue: dto.replacementValue,
      operator: dto.operator as any,
      value: dto.value
    };
  }
  private static toTransformationDto(domain: QueryTransformation): QueryTransformationDto {
    return {
      id: domain.id,
      type: domain.type,
      fieldId: domain.fieldId,
      outputLabel: domain.outputLabel,
      formula: domain.formula,
      format: domain.format,
      replacementValue: domain.replacementValue,
      operator: domain.operator,
      value: domain.value
    };
  }
}
